### DISCUSSION TITLE
getServersideProps first page load only

### DISCUSSION THREAD

**Discussion by alex-statsig** [vercel/next.js#37055] [Help] [answered]
Is there a getServersideProps for SSRing the first page load only (not client transitions)?

I want to load some basic data for my webpage on the first request, but after that treat it as a SPA which does not need to hit the server for routing. The issue is having getServersideProps on every route (to SSR the first page load, and avoid an extra round trip for some basic initial required data) causes a full server round trip on every navigation.

I've seen that a custom Document's getInitialProps may behave like this, but that seems hacky (plus the docs say to avoid using it to prepare for React 18).

**Comment by icyJoseph** [comment]
Have you looked into using `shallow` when navigating? 

> https://nextjs.org/docs/routing/shallow-routing

It has limitations, like going from `/` to `/blog` **will still trigger the navigation**, but updating a query, or staying on the same dynamic route, like navigating between `/posts/[post]` using shallow won't trigger `GSSP` on route change.

## Alternative

> This is slightly hacky, and I might have said something wrong, apologies for that, I checked my statements, a couple of times, and found no lies, but who knows. Writing documentation is hard.

Perhaps you could use `getInitialProps` on `_app.tsx`, but that'll disable optimisation some pages. Only those with `getStaticProps` are optimised.

Inside `getInitialProps` you can check if `ctx.req` is undefined, which means it is running on the client, in such a case return an object with `{ fromServer: false, data: null}`

Otherwise, it is running on the server, so you can return, `{ fromServer: true, data: {foo: 'bar'} }`, then you can use this data as props to a Context Provider, which checks if `fromServer` is true, if so, updates the data, otherwise, skips the update.

### Problems

Again, this is kinda not a good idea, because it will disable automatic optimisation, and for optimised pages it'll potentially show stale data.

Pages that define `getStaticProps` are still optimised, but in that case,  these will have build time `initialData`. So when a user navigates directly to them, they'll show this build time data.

On SPA navigation to these optimised pages, you'll need to do other checks to see if you really want update the state with the `initialData` from build time, or if you want to keep the runtime data. 

Another way of saying this, is that `getInitialProps` won't be executed at runtime for pages that used `getStaticProps`.

If you need to fetch data based on something from the incoming request, then you probably do want to opt out from `getStaticProps`, so the above won't be a problem.

One additional thing is that, if a page uses `getServerSideProps`, then `getInitialProps` from `_app`, runs on the server. 

### Implementation

Proceed with caution...

```tsx
/* ❗❗❗ 
Warning: You have opted-out of Automatic Static Optimization due to `getInitialProps` in `pages/_app`. 
This does not opt-out pages with `getStaticProps`
❗❗❗*/
import type { AppContext, AppProps } from "next/app";
import { useEffect, useState } from "react";

function MyApp({
  Component,
  pageProps,
  fromServer,
  initialData,
}: AppProps & { fromServer: boolean; initialData: { foo: string } }) {
  const [data, setData] = useState(initialData);

  useEffect(() => {
    if (fromServer) {
      setData(initialData);
    }
  }, [fromServer, initialData]);

  return <Component {...pageProps} foo={data.foo} />;
}

MyApp.getInitialProps = async ({ ctx }: AppContext) => {
  if (Boolean(ctx.req)) {
    return { fromServer: true, initialData: { foo: "bar" } };
  }
  return { fromServer: false, initialData: null };
};

export default MyApp;
```

With React Dev tools you'll see that the App wrapper, has `initialData` on first load, set to some value (from the server), and then when you navigate, it becomes `null`, but it doesn't update the `useState` hook value.

**Reply by alex-statsig** [reply]
Tried playing around with this, but my imports for getInitialProps can't be imported on the client. Maybe there's a way to hack around that, but like you said, seems a bit sketchy

**Comment by frysztak** [comment, marked as answer]
there's no built-in way of doing that, much to community's dismay. related RFC: https://github.com/vercel/next.js/discussions/29117

**Reply by alex-statsig** [reply]
Thanks for this information, not sure why I was unable to find that discussion. Appreciate it! Hopefully support will be added eventually
